src/Orleans/Providers/IEventStorageProvider.cs

```diff
@@ -0,0 +1,117 @@
+using System.Threading.Tasks;
+using Orleans.Providers;
+using Orleans.Runtime;
+using System.Collections.Generic;
+
+namespace Orleans.Storage
+{
+    /// <summary>
+    /// Interface to be implemented for a event storage provider able to read from and append to streams
+    /// representing the state of Journaled Grains
+    /// </summary>
+    public interface IEventStorageProvider : IProvider, IEventStorage
+    {
+        /// <summary>Logger used by this event storage provider instance.</summary>
+        /// <returns>Reference to the Logger object used by this provider.</returns>
+        /// <seealso cref="Logger"/>
+        Logger Log { get; }
+
+        /// <summary>
+        /// The provider can specify a default stream name to use for a given grain.
+        /// </summary>
+        /// <param name="grainType">The type of the grain</param>
+        /// <param name="grainReference">The grain reference for the grain</param>
+        /// <returns></returns>
+        string DefaultStreamName(Type grainType, GrainReference grainReference);
+    }
+
+    /// <summary>
+    /// The event store interface, as implemented by event storage providers
+    /// </summary>
+    /// 
+    public interface IEventStorage
+    {
+        /// <summary>
+        /// Returns a handle to an event stream that can be used to perform operations on it.
+        /// </summary>
+        /// <param name="streamName">The name of the stream</param>
+        /// <returns>A handle for performing operations on the stream</returns>
+        IEventStreamHandle GetEventStreamHandle(string streamName);
+    }
+
+    /// <summary>
+    /// The event stream interface, as implemented by event storage providers. 
+    /// Should be disposed when no longer needed.
+    /// </summary>
+    public interface IEventStreamHandle : IDisposable
+    {
+        /// <summary>
+        /// The name of this event stream.
+        /// </summary>
+        string StreamName { get; }
+
+        /// <summary>
+        /// Returns the number of events in the event stream, which corresponds to the version number.
+        /// Returns zero for a not-yet-created stream or for a deleted stream.
+        /// </summary>
+        /// <returns>a nonnegative integer representing the number of events in the stream.</returns>
+        Task<int> GetVersion();
+
+        /// <summary>
+        /// Loads a sequence of events from the event stream.
+        /// </summary>
+        /// <typeparam name="E">The base class for the events</typeparam>
+        /// <param name="startAtVersion">The version (stream position) at which to start</param>
+        /// <param name="endAtVersion">The version (stream position) at which to end, or null to return up and including the most recent</param>
+        /// <returns>A <see cref="EventStreamSegment{E}"/> structure containing a subrange of the event stream.</returns>
+        Task<EventStreamSegment<E>> Load<E>(int startAtVersion = 0, int? endAtVersion = null);
```

Comment:
To add to this, there's a comment above:
> having the type parameter E around during deserialization might be quite helpful at some point (e.g. we can deserialize JSON without having to store the type name)

Unless I'm mistaken, the runtime type of each event will almost always be a descendant of `E` and not `E` itself. For example, `E` might be something like `MyBaseEventType`, but the runtime type will be a collection with mixed descendants like `UserAccountCreated`, `NewMessage`, `UserBannedEvent`.

I would recommend using `object` at this level, since there's no strong need for a concrete event type. Alternatively, I recommend using an interface, eg `IEvent`, which has a `Guid Id` property so that no projection to `IEnumerable<KeyValuePair<Guid, TEvent>>` needs to be performed and we can simply pass `ICollection<IEvent>` instead.